### PR TITLE
Better optimization for mempool new peak

### PR DESCRIPTION
**Pull request by Yostra**

### CODE REVIEW

**Review by mariano54 — APPROVED** [review]

File: chia/full_node/mempool_manager.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ def __init__(self, coin_store: CoinStore, consensus_constants: ConsensusConstant
         self.mempool_max_total_cost = int(self.constants.MAX_BLOCK_COST_CLVM * self.constants.MEMPOOL_BLOCK_BUFFER)
 
         # Transactions that were unable to enter mempool, used for retry. (they were invalid)
-        self.potential_cache = PendingTxCache(self.constants.MAX_BLOCK_COST_CLVM * 5)
+        self.potential_cache = PendingTxCache(self.constants.MAX_BLOCK_COST_CLVM * 1)
```
Comment (aqk) [inline chia/full_node/mempool_manager.py:97]: Does this make the pending cache smaller?

File: chia/full_node/mempool_manager.py [hunk 2]
```diff
@@ -517,31 +517,16 @@ async def new_peak(
...
-                        break
-                if not failed:
-                    self.mempool.add_to_pool(item)
-                else:
-                    # If the spend bundle was confirmed or conflicting (can no longer be in mempool), it won't be
-                    # successfully added to the new mempool. In this case, remove it from seen, so in the case of a
-                    # reorg, it can be resubmitted
+                if coin_record.name in self.mempool.removals:
+                    item = self.mempool.removals[coin_record.name]
+                    self.mempool.remove_from_pool(item)
```
Comment (aqk) [inline chia/full_node/mempool_manager.py:524]: Can anything else access this concurrently while we are updating it?

File: chia/full_node/mempool_manager.py [hunk 3]
```diff
@@ -517,31 +517,16 @@ async def new_peak(
...
-                    if removed_coin.name() in changed_coins_set:
-                        failed = True
-                        break
-                if not failed:
-                    self.mempool.add_to_pool(item)
-                else:
-                    # If the spend bundle was confirmed or conflicting (can no longer be in mempool), it won't be
-                    # successfully added to the new mempool. In this case, remove it from seen, so in the case of a
-                    # reorg, it can be resubmitted
+                if coin_record.name in self.mempool.removals:
```
Comment (arvidn) [inline chia/full_node/mempool_manager.py:522]: wouldn't it be simpler and more straight-forward to just check `coin_record.spent` here?

Comment (arvidn) [reply chia/full_node/mempool_manager.py:522]: oh, I don't think I know what `self.mempool.removals` is..

**Comment by arvidn** [comment]
it seems to me that the optimized path is missing the feature of evicting spend-bundles that were not themselves confirmed in the new block, but were invalidated because of it. For example, one of their inputs were spent.

Do we have tests for those cases?

File: chia/full_node/mempool_manager.py (hunk 1 above)
Comment (Yostra) [reply chia/full_node/mempool_manager.py:97]: Yes, that cache is used only for items that fail due to time locks. It's intended for 1-off block misses mostly. (doesn't make sense to have anything larger than that)

File: chia/full_node/mempool_manager.py (hunk 2 above)
Comment (Yostra) [reply chia/full_node/mempool_manager.py:524]: No, everything is done under that asyncio lock

File: chia/full_node/mempool_manager.py (hunk 3 above)
Comment (Yostra) [reply chia/full_node/mempool_manager.py:522]: mempool.removal are {coin: Mempool item} mapping of things already inside the mempool.

**Comment by mariano54** [comment]
> it seems to me that the optimized path is missing the feature of evicting spend-bundles that were not themselves confirmed in the new block, but were invalidated because of it. For example, one of their inputs were spent.
> 
> Do we have tests for those cases?

There is no test for this afaik. That is a good idea. However, that is covered by this PR, because each of the inputs in that spend bundle are an entry in the `self.mempool.removals`, so if any of those coins are spent, we will remove the whole bundle (mempool item).

**Comment by arvidn** [comment]
@Yostra could you please add the tests mentioned above? and any other regression test cases you can think of that aren't already covered.

**Pull request merged**